File: src/2025-02-24-deepclaude-rust/go/exercises/practice/tree-building/tree_building.go
package tree

import (
	"fmt"
	"sort"
)

type Record struct {
	ID     int
	Parent int
	// feel free to add fields as you see fit
}

type Node struct {
	ID       int
	Children []*Node
	// feel free to add fields as you see fit
}

func Build(records []Record) (*Node, error) {
	if len(records) == 0 {
		return nil, nil
	}

	// Sort records by ID
	sort.Slice(records, func(i, j int) bool {
		return records[i].ID < records[j].ID
	})

	// Validate root node
	if records[0].ID != 0 || records[0].Parent != 0 {
		return nil, fmt.Errorf("invalid root node")
	}

	// Validate all IDs are contiguous and parent IDs are valid
	for i, r := range records {
		if r.ID != i {
			return nil, fmt.Errorf("non-contiguous or invalid ID %d", r.ID)
		}
		if r.ID != 0 && r.Parent >= r.ID {
			return nil, fmt.Errorf("invalid parent ID %d for node %d", r.Parent, r.ID)
		}
	}

	// Build tree using a map for quick lookups
	nodes := make(map[int]*Node)
	root := &Node{ID: 0}
	nodes[0] = root

	for i := 1; i < len(records); i++ {
		r := records[i]
		parent, exists := nodes[r.Parent]
		if !exists {
			return nil, fmt.Errorf("parent %d not found for node %d", r.Parent, r.ID)
		}

		node := &Node{ID: r.ID}
		parent.Children = append(parent.Children, node)
		nodes[r.ID] = node
	}

	return root, nil
}

File: src/2025-03-16-12-54-49--a-deepclaude-full-architect-python-0316-2/python/exercises/practice/tree-building/tree_building.py
class Record:
    def __init__(self, record_id, parent_id):
        self.record_id = record_id
        self.parent_id = parent_id


class Node:
    def __init__(self, node_id):
        self.node_id = node_id
        self.children = []


def BuildTree(records):
    """
    Build a tree structure from records containing IDs and parent IDs.
    
    Args:
        records: List of Record objects with record_id and parent_id attributes
        
    Returns:
        The root Node of the tree, or None if records is empty
    
    Raises:
        ValueError: If the records do not form a valid tree
    """
    if not records:
        return None
    
    # Validate records
    records.sort(key=lambda x: x.record_id)
    record_ids = [record.record_id for record in records]
    
    # Check if IDs are continuous and start from 0
    if record_ids != list(range(len(records))):
        raise ValueError("Record id is invalid or out of order.")
    
    # Create all nodes
    nodes = {record.record_id: Node(record.record_id) for record in records}
    
    # Connect nodes based on parent-child relationships
    for record in records:
        # Root validation
        if record.record_id == 0:
            if record.parent_id != 0:
                raise ValueError("Node parent_id should be smaller than it's record_id.")
        # Non-root validation
        else:
            if record.parent_id == record.record_id:
                raise ValueError("Only root should have equal record and parent id.")
            elif record.parent_id > record.record_id:
                raise ValueError("Node parent_id should be smaller than it's record_id.")
            
            # Add this node as a child to its parent
            parent_node = nodes[record.parent_id]
            parent_node.children.append(nodes[record.record_id])
    
    return nodes[0] if nodes else None
